starlite/datastructures/state.py

```diff
@@ -12,11 +25,135 @@ class State(MutableMapping[str, Any]):
 
     _state: Dict[str, Any]
 
-    def __init__(self, state: Optional[Dict[str, Any]] = None) -> None:
-        """Initialize `State` with an optional value.
+    def __init__(self, state: Union["ImmutableState", Dict[str, Any], Iterable[Tuple[str, Any]]]) -> None:
+        """Initialize an `ImmutableState` instance.
+
+        Args:
+             state: An object to initialize the state from. Can be a dict, an instance of 'ImmutableState', or a tuple of key value paris.
+        """
+
+        if isinstance(state, ImmutableState):
+            state = state._state
+
+        if not isinstance(state, dict) and isinstance(state, Iterable):
+            state = dict(state)
+
+        super().__setattr__("_state", state if state is not None else {})
+
+    def __bool__(self) -> bool:
+        """Return a boolean indicating whether the wrapped dict instance has values."""
+        return bool(self._state)
+
+    def __getitem__(self, key: str) -> Any:
+        """Get the value for the corresponding key from the wrapped state object using subscription notation.
+
+        Args:
+            key: Key to access.
+
+        Raises:
+            KeyError
+
+        Returns:
+            A value from the wrapped state instance.
+        """
+        return self._state[key]
+
+    def __iter__(self) -> Iterator[str]:
+        """Return an iterator iterating the wrapped state dict.
+
+        Returns:
+            An iterator of strings
+        """
+        return iter(self._state)
+
+    def __len__(self) -> int:
+        """Return length of the wrapped state dict.
+
+        Returns:
+            An integer
+        """
+        return len(self._state)
+
+    def __getattr__(self, key: str) -> Any:
+        """Get the value for the corresponding key from the wrapped state object using attribute notation.
 
         Args:
-             state: An optional string keyed dict for the initial state.
+            key: Key to retrieve
+
+        Raises:
+            AttributeError: if the given attribute is not set.
+
+        Returns:
+            The retrieved value
+        """
+        try:
+            return self._state[key]
+        except KeyError as e:
+            raise AttributeError from e
+
+    def __copy__(self) -> "ImmutableState":
+        """Return a shallow copy of the given state object.
+
+        Customizes how the builtin "copy" function will work.
+        """
+        return self.__class__(copy(self._state))
+
+    def mutable_copy(self) -> "State":
+        """Return a mutable copy of the state object.
+
+        Returns:
+            A `State`
+        """
+        return State(self._state)
```

Comment:
I'm aware. I thought of actually making this a view rather than a frozen object. It's still immutable. Either way there are cases where users are going to be unhappy. 

What do you think?

Reply:
Hmm. I'm not sure. The current solution is neither nor, which I think is the worst of both worlds :D

So either it should be truly frozen, which means creating a deep copy of the mutable objects it's based on, or it should be a proper view and declared as such (i.e. being named `StateView`).
I have a slight preference for the former (since then it's truly immutable and free from side effects) but maybe we should gather more opinions on this?

Reply:
nah, lets go with the immutable version

Reply:
Okay so you wanna update this to complete immutability?